src/server/acl/acl_family.cc

```diff
@@ -468,6 +469,34 @@ void AclFamily::GetUser(CmdArgList args, ConnectionContext* cntx) {
   (*cntx)->SendSimpleString(acl);
 }
 
+void AclFamily::GenPass(CmdArgList args, ConnectionContext* cntx) {
+  if (args.length() > 1) {
+    (*cntx)->SendError(facade::UnknownSubCmd("GENPASS", "ACL"));
+    return;
+  }
+  uint32_t random_bits = 256;
+  if (args.length() == 1) {
+    auto requested_bits = facade::ArgS(args, 0);
+
+    if (!absl::SimpleAtoi(requested_bits, &random_bits) || random_bits == 0 || random_bits > 4096) {
+      return (*cntx)->SendError(
+          "ACL GENPASS argument must be the number of bits for the output password, a positive "
+          "number up to 4096");
+    }
+  }
+  auto rbe = ServerState::tlocal()->random_bits_engine;
+  std::vector<unsigned char> data((random_bits + 3) / 4);
+  std::generate(begin(data), end(data), std::ref(*rbe));
+
+  const std::string hex_chars = "0123456789abcdef";
+  std::string response;
+  for (unsigned char c : data) {
+    response += hex_chars[c & 0x0F];
+  }
+
```

Comment:
small update, I edited the last step and removed the chore with the module since `number + 3 / 4` covers it :)

Reply:
@kostasrim, first of all thanks for the thorough and speedy feedback!

I don't quite follow the `constexpr random_bytes_produced = sizeof(decltype(urandom.max()));` explanation.
I think the point is that the number of bytes returned can be platform-dependent and this piece of code calculates how many bytes this abstraction returns? 

Or at least for me the important part where we actually generate the random bytes is missing. I'm also kind of wondering if this C++ abstraction makes sense at all in this case, when we just to read a predetermined number of bytes. Sounds like it would be a lot more straight-forward just to do that explicitly :/ (open /dev/urandom, read n bytes).

Moreover, the way I implemented it in the first place was to match Redis's behavior:
> The password generation is also very cheap because *we don't really ask /dev/urandom for bits at every execution**. At startup Redis creates a seed using /dev/urandom, then it will use SHA256 in counter mode, with HMAC-SHA256(seed,counter) as primitive, in order to create more random bytes as needed. This means that the application developer should be feel free to abuse ACL GENPASS to create as many secure pseudorandom strings as needed.


From what I understand, you suggested that I remove the the random engine and read from `/dev/urandom` on every invocation. My idea was that this PRNG would be seeded once (for each thread) and the resulting stream could be reused. However, I submitted this PR perhaps too hastily, and did not check if this is how it actually works.


So, I have two concerns with my own and the proposed solution:
Firstly, this would make the `ACL GENPASS` command definitely a lot slower and spamming this command could result in reduced entropy as well.

Secondly, I'm not sure if this construction is suitable for a secure PRNG. Perhaps it would be safer to employ the same solution as Redis with its AES CTR? In that case I believe each thread should still have its own counter, read its own seed from urandom etc.

Reply:
I do not think we should be creative here unless we understand in security algorithms. Security and simplicity have higher precedence here over performance and if implementing what Redis did achieve these results we could do the same.

Reply:
@UkuLoskit hello hello hello :)

> I don't quite follow the constexpr random_bytes_produced = sizeof(decltype(urandom.max())); explanation.
I think the point is that the number of bytes returned can be platform-dependent and this piece of code calculates how many bytes this abstraction returns?

Yes it's the size of the type the function returns. This should be `4` on all platforms but for sanity sake I like this chore because the standard doesn't mandate what the exact type is.

>Or at least for me the important part where we actually generate the random bytes is missing. I'm also kind of wondering if this C++ abstraction makes sense at all in this case, when we just to read a predetermined number of bytes. Sounds like it would be a lot more straight-forward just to do that explicitly :/ (open /dev/urandom, read n bytes)

Nah, the C++ wraps it for you, including the error handling etc. There is no reason to open this file and read n bytes. That's why you have abstractions, to do it for you.

> Moreover, the way I implemented it in the first place was to match Redis's behavior:

I did not see any HMAC-SHA25 in your PR?
In fact, we can match `redis` behaviour. Instead, of generating a random number each time from `urandom` you can use it as a seed once and then use HMAC (see redis/util.c:860 getRandomBytes) 

> ACL GENPASS command definitely a lot slower 
This is not in the hot path, we can take a hit here -- which I doubt will be measurable

>  spamming this command could result in reduced entropy as well.
Generally speaking, I think `urandom` covers entropy well from its pool

> In that case I believe each thread should still have its own counter, read its own seed from urandom etc.
Atomic would do here, no need for complicated logic. This is not in the hot path, even for other ACL commands we use locks...

> I'm not sure if this construction is suitable for a secure PRNG. 
From google (as I said I am not an expert in security), urandom should be pretty secure :) 

I agree with @romange that we should be as close to redis as possible so if you are ok with it, you can take the HMAC route (I would be also happy with urandom).

Reply:
@kostasrim and @romange, as my work day is starting now, I tried to submit my changes before that.
I changed it to the inline version that @kostasrim suggested in the first place, perhaps it's for the best not to overengineer the initial solution too much and only revisit it in the future if there really are any performance concerns. From my investigation using `strace` ti seems that this actually calls `getrandom` under the hood and asks for far fewer bytes.

finally, @kostasrim , I did not know how apply your constexpr solution, so suggestions welcome for this.